File: packages/core/src/heartbeat-monitor.ts
import type { TaskEngine } from './engine.js'
import type { ShortTermStore, DisconnectPolicy, Worker, TaskStatus } from './types.js'
import type { WorkerManager } from './worker-manager.js'
import { canTransition } from './state-machine.js'

export interface HeartbeatMonitorOptions {
  workerManager: WorkerManager
  engine: TaskEngine
  shortTermStore: ShortTermStore
  checkIntervalMs?: number          // default 30_000
  heartbeatTimeoutMs?: number       // default 90_000
  defaultDisconnectPolicy?: DisconnectPolicy  // default 'reassign'
  disconnectGraceMs?: number        // default 30_000
}

export class HeartbeatMonitor {
  private workerManager: WorkerManager
  private engine: TaskEngine
  private shortTermStore: ShortTermStore
  private checkIntervalMs: number
  private heartbeatTimeoutMs: number
  private defaultDisconnectPolicy: DisconnectPolicy
  private disconnectGraceMs: number
  private timer?: ReturnType<typeof setInterval>
  private graceTimers = new Map<string, ReturnType<typeof setTimeout>>()

  constructor(opts: HeartbeatMonitorOptions) {
    this.workerManager = opts.workerManager
    this.engine = opts.engine
    this.shortTermStore = opts.shortTermStore
    this.checkIntervalMs = opts.checkIntervalMs ?? 30_000
    this.heartbeatTimeoutMs = opts.heartbeatTimeoutMs ?? 90_000
    this.defaultDisconnectPolicy = opts.defaultDisconnectPolicy ?? 'reassign'
    this.disconnectGraceMs = opts.disconnectGraceMs ?? 30_000
  }

  start(): void {
    this.timer = setInterval(() => this.tick().catch(() => {}), this.checkIntervalMs)
  }

  stop(): void {
    if (this.timer) clearInterval(this.timer)
    for (const t of this.graceTimers.values()) clearTimeout(t)
    this.graceTimers.clear()
  }

  async tick(): Promise<void> {
    // List all non-offline workers
    const workers = await this.shortTermStore.listWorkers({ status: ['idle', 'busy', 'draining'] })
    const now = Date.now()

    for (const worker of workers) {
      if (now - worker.lastHeartbeatAt > this.heartbeatTimeoutMs) {
        // Already processing this worker
        if (this.graceTimers.has(worker.id)) continue

        await this._handleTimeout(worker)
      }
    }
  }

  /** Cancel grace timer if worker comes back (call from heartbeat handler) */
  cancelGrace(workerId: string): void {
    const timer = this.graceTimers.get(workerId)
    if (timer) {
      clearTimeout(timer)
      this.graceTimers.delete(workerId)
    }
  }

  private async _handleTimeout(worker: Worker): Promise<void> {
    // Mark worker offline
    worker.status = 'offline'
    await this.shortTermStore.saveWorker(worker)

    // Get all assignments for this worker
    const assignments = await this.workerManager.getWorkerTasks(worker.id)

    // Determine policy per task (task-level override or default)
    for (const assignment of assignments) {
      const task = await this.engine.getTask(assignment.taskId)
      if (!task) continue

      const policy = task.disconnectPolicy ?? this.defaultDisconnectPolicy

      switch (policy) {
        case 'fail':
          try {
            await this.engine.transitionTask(assignment.taskId, 'failed', {
              error: { code: 'WORKER_DISCONNECT', message: `Worker ${worker.id} disconnected (heartbeat timeout)` },
            })
          } catch { /* task may already be terminal */ }
          await this.workerManager.releaseTask(assignment.taskId)
          break

        case 'mark':
          // Just mark offline, leave tasks as-is for external intervention
          break

        case 'reassign':
          this._startGraceTimer(worker.id, assignment.taskId)
          break
      }
    }
  }

  /**
   * Transition a task back to pending through valid state machine transitions.
   * running -> paused -> assigned -> pending
   * assigned -> pending
   * Other states -> try direct pending transition
   */
  private async _transitionToPending(taskId: string): Promise<void> {
    const task = await this.engine.getTask(taskId)
    if (!task) return

    const status = task.status

    // Build a path from current status to pending
    const path = this._findPathToPending(status)
    for (const step of path) {
      try {
        await this.engine.transitionTask(taskId, step)
      } catch {
        // Transition may fail if task was changed externally
        return
      }
    }
  }

  private _findPathToPending(from: TaskStatus): TaskStatus[] {
    if (from === 'pending') return []
    if (canTransition(from, 'pending')) return ['pending']

    // running -> paused -> assigned -> pending
    if (from === 'running') return ['paused', 'assigned', 'pending']
    // paused -> assigned -> pending
    if (from === 'paused') return ['assigned', 'pending']
    // blocked -> assigned -> pending
    if (from === 'blocked') return ['assigned', 'pending']

    return []
  }

  private _startGraceTimer(workerId: string, taskId: string): void {
    // Don't duplicate grace timers for the same worker
    if (this.graceTimers.has(workerId)) return

    const timer = setTimeout(async () => {
      this.graceTimers.delete(workerId)

      // Check if worker came back
      const worker = await this.shortTermStore.getWorker(workerId)
      if (worker && worker.status !== 'offline') return

      // Reassign: revert task to pending through valid transitions
      await this._transitionToPending(taskId)
      await this.workerManager.releaseTask(taskId)
    }, this.disconnectGraceMs)

    this.graceTimers.set(workerId, timer)
  }
}

File: rust/taskcast-core/src/heartbeat_monitor.rs
use std::collections::HashSet;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use tokio::sync::RwLock;
use tokio::task::JoinHandle;
use tokio::time::{sleep, interval, Duration};

use crate::engine::{TaskEngine, TransitionPayload};
use crate::types::{
    DisconnectPolicy, ShortTermStore, TaskError, TaskStatus, WorkerFilter, WorkerStatus,
};
use crate::worker_manager::WorkerManager;

// ─── Options ─────────────────────────────────────────────────────────────────

pub struct HeartbeatMonitorOptions {
    pub worker_manager: Arc<WorkerManager>,
    pub engine: Arc<TaskEngine>,
    pub short_term_store: Arc<dyn ShortTermStore>,
    /// How often to check heartbeats, in milliseconds. Default: 30_000.
    pub check_interval_ms: u64,
    /// How long before a worker is considered timed out, in milliseconds. Default: 90_000.
    pub heartbeat_timeout_ms: u64,
    /// Default disconnect policy when a task has no explicit policy. Default: Reassign.
    pub default_disconnect_policy: DisconnectPolicy,
    /// Grace period before reassigning, in milliseconds. Default: 30_000.
    pub disconnect_grace_ms: u64,
}

// ─── HeartbeatMonitor ───────────────────────────────────────────────────────

pub struct HeartbeatMonitor {
    worker_manager: Arc<WorkerManager>,
    engine: Arc<TaskEngine>,
    short_term_store: Arc<dyn ShortTermStore>,
    check_interval_ms: u64,
    heartbeat_timeout_ms: u64,
    default_disconnect_policy: DisconnectPolicy,
    disconnect_grace_ms: u64,
    handle: Option<JoinHandle<()>>,
    /// Set of worker IDs currently in a grace period (pending reassignment).
    grace_workers: Arc<RwLock<HashSet<String>>>,
}

impl HeartbeatMonitor {
    pub fn new(opts: HeartbeatMonitorOptions) -> Self {
        Self {
            worker_manager: opts.worker_manager,
            engine: opts.engine,
            short_term_store: opts.short_term_store,
            check_interval_ms: opts.check_interval_ms,
            heartbeat_timeout_ms: opts.heartbeat_timeout_ms,
            default_disconnect_policy: opts.default_disconnect_policy,
            disconnect_grace_ms: opts.disconnect_grace_ms,
            handle: None,
            grace_workers: Arc::new(RwLock::new(HashSet::new())),
        }
    }

    pub fn start(&mut self) {
        let worker_manager = self.worker_manager.clone();
        let engine = self.engine.clone();
        let store = self.short_term_store.clone();
        let interval_ms = self.check_interval_ms;
        let timeout_ms = self.heartbeat_timeout_ms;
        let default_policy = self.default_disconnect_policy.clone();
        let grace_ms = self.disconnect_grace_ms;
        let grace_workers = self.grace_workers.clone();

        self.handle = Some(tokio::spawn(async move {
            let mut ticker = interval(Duration::from_millis(interval_ms));
            loop {
                ticker.tick().await;
                let _ = Self::tick_inner(
                    &worker_manager,
                    &engine,
                    &store,
                    timeout_ms,
                    &default_policy,
                    grace_ms,
                    &grace_workers,
                )
                .await;
            }
        }));
    }

    pub fn stop(&mut self) {
        if let Some(handle) = self.handle.take() {
            handle.abort();
        }
    }

    /// Public tick for testing — runs one check cycle immediately.
    pub async fn tick(&self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        Self::tick_inner(
            &self.worker_manager,
            &self.engine,
            &self.short_term_store,
            self.heartbeat_timeout_ms,
            &self.default_disconnect_policy,
            self.disconnect_grace_ms,
            &self.grace_workers,
        )
        .await
    }

    async fn tick_inner(
        worker_manager: &Arc<WorkerManager>,
        engine: &Arc<TaskEngine>,
        store: &Arc<dyn ShortTermStore>,
        timeout_ms: u64,
        default_policy: &DisconnectPolicy,
        grace_ms: u64,
        grace_workers: &Arc<RwLock<HashSet<String>>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let workers = store
            .list_workers(Some(WorkerFilter {
                status: Some(vec![
                    WorkerStatus::Idle,
                    WorkerStatus::Busy,
                    WorkerStatus::Draining,
                ]),
                connection_mode: None,
            }))
            .await?;

        let now = now_millis();

        for worker in workers {
            if now - worker.last_heartbeat_at > timeout_ms as f64 {
                // Skip workers already in grace period
                if grace_workers.read().await.contains(&worker.id) {
                    continue;
                }

                Self::handle_timeout(
                    worker_manager,
                    engine,
                    store,
                    &worker.id,
                    default_policy,
                    grace_ms,
                    grace_workers,
                )
                .await?;
            }
        }

        Ok(())
    }

    async fn handle_timeout(
        worker_manager: &Arc<WorkerManager>,
        engine: &Arc<TaskEngine>,
        store: &Arc<dyn ShortTermStore>,
        worker_id: &str,
        default_policy: &DisconnectPolicy,
        grace_ms: u64,
        grace_workers: &Arc<RwLock<HashSet<String>>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        // Mark worker offline
        let worker = store.get_worker(worker_id).await?;
        let Some(mut worker) = worker else {
            return Ok(());
        };
        worker.status = WorkerStatus::Offline;
        store.save_worker(worker).await?;

        // Get all current assignments for this worker
        let assignments = worker_manager.get_worker_tasks(worker_id).await?;

        for assignment in assignments {
            let task = engine.get_task(&assignment.task_id).await?;
            let Some(task) = task else { continue };

            let policy = task
                .disconnect_policy
                .clone()
                .unwrap_or_else(|| default_policy.clone());

            match policy {
                DisconnectPolicy::Fail => {
                    let _ = engine
                        .transition_task(
                            &assignment.task_id,
                            TaskStatus::Failed,
                            Some(TransitionPayload {
                                error: Some(TaskError {
                                    code: Some("WORKER_DISCONNECT".to_string()),
                                    message: format!(
                                        "Worker {} disconnected (heartbeat timeout)",
                                        worker_id
                                    ),
                                    details: None,
                                }),
                                ..Default::default()
                            }),
                        )
                        .await;
                    let _ = worker_manager.release_task(&assignment.task_id).await;
                }
                DisconnectPolicy::Mark => {
                    // Worker is already marked offline above; nothing else to do.
                }
                DisconnectPolicy::Reassign => {
                    // Start grace timer
                    grace_workers.write().await.insert(worker_id.to_string());

                    let wm = worker_manager.clone();
                    let eng = engine.clone();
                    let store_clone = store.clone();
                    let wid = worker_id.to_string();
                    let tid = assignment.task_id.clone();
                    let grace_set = grace_workers.clone();

                    tokio::spawn(async move {
                        sleep(Duration::from_millis(grace_ms)).await;
                        grace_set.write().await.remove(&wid);

                        // Check if worker came back during grace period
                        if let Ok(Some(w)) = store_clone.get_worker(&wid).await {
                            if w.status != WorkerStatus::Offline {
                                return;
                            }
                        }

                        // Worker is still offline — reassign the task
                        let _ = eng
                            .transition_task(&tid, TaskStatus::Pending, None)
                            .await;
                        let _ = wm.release_task(&tid).await;
                    });
                }
            }
        }

        Ok(())
    }
}

fn now_millis() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("system time before UNIX epoch")
        .as_millis() as f64
}
